galleries/examples/images_contours_and_fields/image_exact_placement.py

```diff
@@ -0,0 +1,131 @@
+axs[0].add_patch(rect)
+
+# %%
+# Note that both images are rendered at a 1:1 ratio, but are made to look almost the
+# same width, despite image B being smaller than image A.
+#
+# We can control the relative sizes using the *width_ratios* argument *if* the images
+# are wider than they are tall and shown side by side, as is the case here.
+
+fig, axs = plt.subplots(1, 2, width_ratios=[300/200, 1], facecolor='aliceblue')
+
+axs[0].imshow(A, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+axs[0].add_patch(rect)
+axs[1].imshow(B, vmin=-1, vmax=1)
+
+# %%
+# Given that the data subsample is in the upper left of the larger image,
+# it might make sense if the top of the smaller Axes aligned with the top of the larger.
+# This can be done manually by using `~.Axes.set_anchor`, and using "NW" (for
+# northwest).
+
+fig, axs = plt.subplots(1, 2, width_ratios=[300/200, 1], facecolor='aliceblue')
+
+axs[0].imshow(A, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+axs[0].add_patch(rect)
+axs[0].set_anchor('NW')
+axs[1].imshow(B, vmin=-1, vmax=1)
+axs[1].set_anchor('NW')
+
+# %%
+# Note that this procedure still leaves large white spaces (that can be trimmed
+# in a final product by ``bbox_inches="tight"`` in `~.Figure.savefig`), and is
+# not very general.  For instance, if the axes had been arranged vertically
+# instead of horizontally, setting the height aspect ratio would not have
+# helped because the axes are wider than they are tall. For more complicated
+# situations it is necessary to place the axes manually.
+#
+# Manual placement
+# ================
+#
+# We can manually place axes when they are created by passing a position to
+# `~.Figure.add_axes`.  This position takes the form ``[left bottom width height]`` and
+# is in units that are a fraction of the figure width and height. Here we decide how
+# large to make the axes based on the size of the images, and add a small buffer of
+# 0.35 inches.  We do all this at 100 dpi.
+
+dpi = 100  # 100 pixels is one inch
+buffer = 0.35 * dpi  # pixels
+left = buffer
+bottom = buffer
+ny, nx = np.shape(A)
+posA = [left, bottom, nx, ny]
+# we know this is tallest, so we can get the fig height:
+figh = bottom + ny + buffer
+
+# place the B axes
+left = left + nx + buffer
+ny, nx = np.shape(B)
+posB = [left, figh - buffer - ny, nx, ny]
+
+# get the fig width
+figw = left + nx + buffer
+
+fig = plt.figure(figsize=(figw / dpi, figh / dpi), facecolor='aliceblue')
+
+ax = fig.add_axes([posA[0] / figw, posA[1] / figh, posA[2] / figw, posA[3] / figh])
+ax.imshow(A, vmin=-1, vmax=1)
+rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
+ax.add_patch(rect)
+
+ax = fig.add_axes([posB[0] / figw, posB[1] / figh, posB[2] / figw, posB[3] / figh])
+ax.imshow(B, vmin=-1, vmax=1)
```

Comment:
Suggestion structuring to make the calculations more understandable:
1. Calculate desired figure height and width in pixel
2. Create figure (needs calculation of figure size in inches)
3. Add axes (needs calculation of dimensions in figure coordinates) 

```suggestion
dpi = 100  # 100 pixels is one inch
margin_px = 0.35 * dpi  # in pixels
ny, nx = np.shape(A)
fig_height_px = ny + 2 * margin
fig_width_px = nx + 2 * margin

# calculate figsize in inches
figsize = (fig_width_px / dpi, fig_height_px / dpi)
fig = plt.figure(figsize=figsize, facecolor='aliceblue')

# add axes in figure coordinates
ax1 = fig.add_axes([
    margin_px / fig_width_px,   # left
    margin_px / fig_height_px,  # bottom
    nx / fig_width_px,          # width
    ny / fig_height_px          # height
])
ax2 = fig.add_axes([
    margin_px / fig_width_px,   # left
    (fig_height_px - margin_px - ny) / fig_height_px,  # bottom
    nx / fig_width_px,          # width
    ny / fig_height_px          # height
])

ax1.imshow(A, vmin=-1, vmax=1)
rect = mpatches.Rectangle((0, 0), 200, 40, linewidth=1, edgecolor='r', facecolor='none')
ax.add_patch(rect)
ax2.imshow(B, vmin=-1, vmax=1)
```

Reply:
A and B are purposefully different widths, so I don't think this is correct.

Reply:
You  are right. The numbers / expressions are not correct. Still the idea (which possibly can be communicated more explicitly) is that you can place everything at exact pixel positions. And to do that, you have to follow the three steps above.

Reply:
Possibly now correct. - I don't have an interpreter at hand, so can't check, but I hope you get the idea.

Reply:
I do get the idea, but not convinced pre-calculating the figure dimensions buys you anything.  Also, imagine we were to add a third or fourth image in there, we would need to change the code in two places.

Just for context, I originally wrote as a for-loop to allow flexibility, but backed off on that as perhaps too magical.

Reply:
> Just for context, I originally wrote as a for-loop to allow flexibility, but backed off on that as perhaps too magical.

👍 Examples must show the pattern, not necessarily a generic solution if that makes the code harder to understand.

---

I had difficulty following your code. Reflecting on it, the reasons were
1. lines 100-115 define a lot of variables, which are sometimes
  a) tersely named (`figh`)
   b) cryptic data structures `posB = [left, figh - buffer - ny, nx, ny]`
   c) a mixture of definitions for later use (e.g. `figh`) and temporary local variables that are reassigned (`left`, `nx`, `ny´)
2. It's unclear that these are all in pixels
3. Transformation calculations to inches (for figsize) and figure units (for add_axes) are hard to understand, because both the input units (see 2.) and the resulting units are not clear.

I tried a more minimal refactor of your code to alleviate these points, but I ended up with almost the same code as I've proposed above 🤷‍♂️.